### PR TITLE
Support getting a shared CUDA tensor on same process

### LINKED ISSUE #7096: invalid device ordinal when using torch.multiprocessing.Queue
```
import torch
import torch.multiprocessing as mp
import numpy.random as npr

mp.set_start_method('spawn')
q=mp.Queue()

for i in range(4):
    a=torch.zeros(npr.randint(1, 3)).cuda()
    q.put(a)

a=q.get() # error cuda runtime error (10) : invalid device ordinal at /pytorch/torch/csrc/generic/StorageSharing.cpp:304 !!!
```

### PR DESCRIPTION
**Pull request by zou3519**
Attempting to fix #7096 

Right now, if one attempts to share a CUDA tensor and get it from a
queue from the same process it was saved in, a crash can occur.

Here is some example code to demonstrate this:

```
import torch
import torch.multiprocessing as mp
import numpy.random as npr

mp.set_start_method('spawn')

q = mp.Queue()
x = torch.randn(1000).cuda()
tensor = torch.ones(10).cuda()

q.put(tensor)
out = q.get()
print(out)
```

One of two behaviors can occur:
- Crashing with an error that some "CUDA arguments are incorrect"
- The `out` tensor is valid but contains incorrect data (zeros instead
of ones).

On master the first behavior occurs but I've noticed the second behavior
happen on previous commits.

This occurs because caching of the shared storages happens differently
depending on the process. Call the process where `tensor` is created
the "originating process". When a process grabs a tensor from the queue
`q`, there are two cases:
1) The process is the "originating process". The cached storage is the
desired storage.
2) The process is not the "originating process". The cached storage (if
it has been cached) is a storage that points to the base of the
allocation. One needs to add the offset to this storage to retrieve the
desired storage.

Case (2) is OK, but the code doesn't handle case (1) right now (the
rebuilding_storage_cuda code attempts to add the offset to the desired
storage, leading to an incorrect storage pointer).

The fix in this PR is to handle case (1) separately from case (2) by:
- Including the PID of the "originating process" when reducing a storage
- Compare the current proc's PID to the PID of the "originating process"
  to determine which case we're in and handle it accordingly.

One alternative solution would be to not have this discrepancy in caching
storages depending on the process, ie, "the cached (CUDA) storage should always
point to the base of the allocation" would be made into an invariant.
I wasn't sure how to work with this because the "originating process"
would need to save a new storage into the cache that points to the base
of the allocation. There's nothing else alive in Python-land that holds
a reference to this new storage so it would be freed immediately when it
goes out of scope.

Another alternative is to just ban this behavior outright, ie, "the
originating process should never be allowed to rebuild a storage that
came from a tensor it created". Saving the pids in the `reduce_storage`
step would then be used for error checking when the storage is being
rebuilt.

cc @colesbury let me know what you think

### CODE REVIEW

File: torch/multiprocessing/reductions.py [hunk 1]
```diff
@@ -88,14 +88,27 @@ def rebuild_storage_filename(cls, manager, handle, size):
     return storage._shared_decref()
 
 
-def rebuild_storage_cuda(cls, device, handle, size, offset, view_size):
+def rebuild_storage_cuda(cls, device, handle, size, offset, view_size, originating_pid):
```
Comment (apaszke) [inline torch/multiprocessing/reductions.py]: Does this handle a situation where process `A` sends a tensor to `B`, and then `B` sends it back to `A`? I think in that case `originating_pid` would point to that of `B`. Additionally, that's not a very robust mechanism, because pids behave effectively like Python's `id` - you're guaranteed that it's unique for as long as a process is alive, but it can be reused once it dies, so you can't be 100% sure if this message was sent by you or not.

Maybe a better fix would be to just always normalize the storages in the cache to point to the beginning of the data block?

**Review by apaszke — CHANGES_REQUESTED** [review]

**Comment by apaszke** [comment]
Sorry I read your PR message only now. I think a simpler way to do that would be to add attributes to storages that are living in the cache. Just take the `_weak_ref` return value, and do `ref.__original = True`. Then, use that in the function that rebuilds the storage to check if you're in the process that has created the storage.

File: torch/multiprocessing/reductions.py (hunk 1 above)
Comment (zou3519) [reply torch/multiprocessing/reductions.py]: I think you're right. I wrote a script to do that: process A sends a tensor to B, then B sends it back to A, and it appears to deadlock...

**Comment by zou3519** [comment]
Thanks @apaszke for the suggestion! I think that should work and it's a lot simpler

**Comment by zou3519** [comment]
Actually, the approach of "saving the storage as-is if we're in the process that created the storage" doesn't completely work. Here's why:
Let's say this is our allocation block:
```
|               [storage1]    [storage2]    |
^
base
```
Now, let's say we share `storage1` and `storage2`, in that order:
- When `storage1` is put into a `mp.Queue`, we cache a reference to it using the `handle` to the entire allocation block. 
- When `storage2` is put into the `mp.Queue`, we cache a reference to it using the `handle` to the entire allocation block, replacing the reference to `storage1`.

Now, let's say another process sends `storage1` back to this process (that created `storage1`). It'll grab the storage for `storage2` and use that instead, which is a problem.

This approach can work if we tweak it so that each process saves all cuda storages that it creates into a separate cache. I'll try doing something like that in another PR.

**Comment by zou3519** [comment]
Closing this in favor of #7204 (it was easier to open a new pr than build on top of this one)

**Comment by apaszke** [comment]
That's true, nice catch @zou3519!
